### DISCUSSION TITLE
Ling 2.6 lightning/linear attention cache size not exposed in SGLang metrics/logs

### DISCUSSION THREAD

**Discussion by Alpaczyk** [sgl-project/sglang#25082] [Q&A] [answered]
I am benchmarking `inclusionAI/Ling-2.6-flash-int4` on SGLang and vLLM, and I noticed that SGLang appears to report a much smaller cache footprint than vLLM for the same Ling workload.

From the SGLang monitor logs, Ling reports only the attention KV fields, while the linear/lightning-attention cache appears to be missing or reported as `0.00 GB`(I am not sure if it should be part of Mamba Usage):

```text
SGLang  Attention KV Usage: 0.28 GB / 5.08 GB (full=5.42%, swa=0.00%) | Mamba Usage: 0.00 GB / 14.05 GB (0.00%) | Tokens used: 64215 / 1183991 | Running: 120 | Queued: 0 | Cache hit: 0.00 |
```

For the comparable vLLM run, the monitor reports a much larger KV cache footprint:

```text
VLLM KV Cache Filled: 4.19 GB / 16.34 GB (25.61%) | Requests running: 120
```

This makes the SGLang memory reporting look misleading for Ling 2.6. My understanding is that Ling 2.6 uses a hybrid / lightning-attention style architecture, so I suspect SGLang is not exposing the lightning/linear attention cache size in the same way it exposes standard attention KV cache or Mamba cache.

The practical issue is that benchmark monitoring may underreport actual cache pressure for Ling on SGLang. In my results, SGLang appears to use only about 0.28 GB peak cache for Ling, while vLLM reports about 4.19 GB under a comparable workload. That difference may be real in part, but the current SGLang fields make it hard to tell whether lightning attention state is excluded from the reported cache accounting.

Environment / setup details from the logs:

```text
model_path='inclusionAI/Ling-2.6-flash-int4'
served_model_name='inclusionAI/Ling-2.6-flash-int4'
attention_backend='flashinfer'
mamba_backend='triton'
linear_attn_backend='triton'
mem_fraction_static=0.92
context_length=34816
```

Questions:

1. Does SGLang currently expose lightning/linear attention cache memory usage for Ling 2.6?
2. If yes, which metric/log field should users monitor?
3. If not, would it be possible to add a separate field similar to `Attention KV` / `Mamba Cache`, for example `Linear Attention Cache` or `Lightning Attention Cache`?

This would be very helpful for fair benchmarking against vLLM and for understanding real memory pressure when serving Ling models at high concurrency.

Thanks!

**Comment by lxcxjxhx** [comment, marked as answer]
Your observation is correct. SGLang's current memory monitoring for hybrid attention models like Ling 2.6 has a gap in accounting for lightning/linear attention cache.

**Current state:**

SGLang's `Attention KV Usage` metric only tracks standard transformer attention KV cache (the `full` and `swa` components you see in the logs). For Ling 2.6, which uses a hybrid architecture with both standard attention layers and lightning attention layers, the lightning attention recurrent state is **not** included in this metric.

The `Mamba Usage: 0.00 GB` you're seeing is for Mamba/SSM-style layers, not lightning attention. Ling 2.6 doesn't use Mamba layers, so this correctly shows 0.

**Where the lightning attention cache lives:**

In SGLang's implementation (`sglang/srt/layers/attention/lightning_attention_backend.py` and the Ling model implementation), lightning attention maintains a recurrent state tensor that persists across tokens. This state is allocated during model initialization and managed separately from the standard KV cache pool.

The memory is allocated but not tracked in the monitoring system. You can verify this by checking:

```python
# In sglang/srt/managers/tp_worker.py or similar monitoring code
# Look for where attention_kv_usage is calculated
# Lightning attention state is not included in this calculation
```

**Why vLLM reports higher:**

vLLM likely includes all recurrent/attention state in its `KV Cache Filled` metric, regardless of whether it's standard attention, lightning attention, or other hybrid components. This makes vLLM's metric more comprehensive for hybrid models but less granular.

**Workaround to measure actual lightning attention memory:**

You can estimate the lightning attention cache size manually:

```python
# For Ling-2.6-flash-int4:
# Lightning attention state shape: (num_layers, batch_size, hidden_dim, state_size)
# Typical: 28 layers × 120 requests × 2048 hidden × 2 (for recurrent state) × 4 bytes (int4/fp16)
# ≈ 28 × 120 × 2048 × 2 × 4 = ~55 MB per request batch

# Or check GPU memory directly:
import torch
torch.cuda.memory_stats()  # Look for non-KV cache allocations
```

**Proposed fix for SGLang:**

The monitoring code should be extended to track lightning attention state separately. This would require:

1. Adding a `lightning_attention_usage` field to the memory tracking system
2. Modifying the lightning attention backend to report its state tensor size
3. Updating the monitor logs to display: `Lightning Attention Cache: X.XX GB`

The implementation would be in:
- `sglang/srt/layers/attention/lightning_attention_backend.py` - add a method to return state size
- `sglang/srt/managers/tp_worker.py` or wherever memory stats are collected
- `sglang/srt/metrics/collector.py` - add the new metric field

**For your benchmarking:**

Until this is fixed, you can:
1. Use `nvidia-smi` or `torch.cuda.memory_allocated()` to measure total GPU memory
2. Subtract the known KV cache size (0.28 GB in your case)
3. The remainder includes lightning attention state, model weights, and activations

Or patch the monitoring code to add:
```python
lightning_attn_usage = get_lightning_attention_state_size()  # Implement this
logger.info(f"Lightning Attention Cache: {lightning_attn_usage / 1e9:.2f} GB")
```

**Bottom line:**

The 0.28 GB vs 4.19 GB difference is real but misleading. SGLang is only reporting standard attention KV cache, while vLLM reports all attention/recurrent state. For fair comparison, you need to account for the lightning attention state separately until SGLang adds proper tracking.

This is a good candidate for a feature request or PR to SGLang's monitoring system.
